axelor-contract/src/main/java/com/axelor/apps/contract/service/ContractServiceImpl.java

```diff
@@ -898,4 +899,51 @@ public void setInitialPriceOnContractLines(Contract contract) {
     }
     contractRepository.save(contract);
   }
+
+  @Override
+  public boolean checkConsumptionLineQuantity(
+      Contract contractCtx, ConsumptionLine consumptionLineCtx, BigDecimal initQt) {
+
+    BigDecimal max = BigDecimal.ZERO;
+    if (!contractCtx.getCurrentContractVersion().getContractLineList().isEmpty()) {
+      List<ContractLine> contractLines =
+          contractCtx.getCurrentContractVersion().getContractLineList().stream()
+              .filter(
+                  cl ->
+                      cl.getIsConsumptionLine()
+                          && Objects.equals(
+                              cl.getProduct().getId(), consumptionLineCtx.getProduct().getId()))
+              .collect(Collectors.toList());
+      if (contractLines.isEmpty()) {
+        return false;
+      }
+      max = contractLines.get(0).getConsumptionMaxQuantity();
+    }
+    BigDecimal sum =
+        contractCtx.getConsumptionLineList().stream()
+            .filter(
+                consumptionLine ->
+                    dateInPeriod(
+                            consumptionLine.getLineDate(),
+                            contractCtx.getInvoicePeriodStartDate(),
+                            contractCtx.getInvoicePeriodEndDate())
+                        && consumptionLine
+                            .getProduct()
+                            .getId()
+                            .equals(consumptionLineCtx.getProduct().getId())
+                        && !consumptionLine.getIsInvoiced())
+            .map(ConsumptionLine::getQty)
+            .reduce(BigDecimal.ZERO, BigDecimal::add);
+
+    sum = sum.subtract(initQt);
+    sum = sum.add(consumptionLineCtx.getQty());
+    return sum.compareTo(max) > 0;
```

Comment:
the case when there is no max quantity defined is not managed

Reply:
there is no need to manage it, since cosumptionMaxQuantity is defined in the domain with min=0, even if it's null and the request is done directly by a direct API call, the getter would return 0 not null, here is the getter:
public BigDecimal getConsumptionMaxQuantity() {
    return consumptionMaxQuantity == null ? BigDecimal.ZERO : consumptionMaxQuantity;
  }

Reply:
I meant functionnaly, we do not want to always have a max quantity